---
title: "On Using the DLOmix service from R"
author:
- name: Christian Panse
  affiliation:
    - &id Functional Genomics Center Zurich - Functional Genomics Center Zurich (FGCZ) - University of Zurich | ETH Zurich, Winterthurerstrasse 190, CH-8057 Zurich, Switzerland
    - Swiss Institute of Bioinformatics (SIB), Quartier Sorge - Batiment Amphipole, CH-1015 Lausanne, Switzerland
  email: cp@fgcz.ethz.ch
package: koina
abstract: |
  This vignette provides ...
output:
  BiocStyle::html_document:
    toc_float: true
bibliography: koina.bib
vignette: >
  %\usepackage[utf8]{inputenc}
  %\VignetteIndexEntry{primer}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
urlcolor: blue
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
knitr::opts_chunk$set(fig.wide = TRUE, fig.retina = 3, error=FALSE)
```


# Initial

```{r load}
lp <- file.path(Sys.getenv("HOME"),
                "src/v2.34.0dev.clients/lib/libgrpcclient.so")
stopifnot(file.exists(lp))

dyn.load(lp)
library(koina)
```

# Hello, `LKEATIQLDELNQK`!


https://www.nature.com/articles/s41592-019-0426-7

```{r hello}
koina::prosit2019Intensity("LKEATIQLDELNQK",
                                    collisionEnergy = 25,
                                    precursorCharge = 2L,
                                    verbose = TRUE,
				    usessl = TRUE,
                                    url = "koina.proteomicsdb.org:443") |> head()
```


```{r helperFunction}
## define some R helper function ...
.plotPrositIntensity <- function(peptide,
                                 precursorCharge = 3,
                                 collisionEnergy = 35,
                                 ...){
  #message(paste("predicting peptide", peptide))
  #peptide <- "LKEATIQLDELNQK"
  #precursorCharge = 3
  #collisionEnergy = 35
  p <- koina::prosit2019Intensity(peptide,
      precursorCharge = precursorCharge,
      collisionEnergy = collisionEnergy,
      usessl = TRUE)[[1]]

  p <- p[!is.na(p$mz),]
  n <- nchar(peptide)
  
  yIons <- protViz::fragmentIon(peptide)[[1]]$y[seq(1, n-1)]
  bIons <- protViz::fragmentIon(peptide)[[1]]$b[seq(1, n-1)]
  names(yIons) <- paste0('y', seq(1, n-1), "(1+)")
  names(bIons) <- paste0('b', seq(1, n-1), "(1+)")

  ## retrieve mZ and intensities 
  #intensity <- p[seq(1, 6 * n)]
  #mZ <- p[175 + seq(1, 6 * n)]
  
  idx <- order(p$mz)
  mZ <- p$mz[idx]
  intensity <- p$intensity[idx]
  
  ## determine b- and y-ion match
  yIdx <- protViz::findNN(yIons, mZ)
  yHit <- abs(mZ[yIdx] - yIons) < 0.1
  
  bIdx <- protViz::findNN(bIons, mZ)
  bHit <- abs(mZ[bIdx] - bIons) < 0.1
  
  plot(intensity ~ mZ, type = 'h', ...)
  
  points(mZ[yIdx[yHit]], intensity[yIdx[yHit]],
         col = 'red', type = 'h', lwd = 2.0)
  
  points(mZ[bIdx[yHit]], intensity[bIdx[bHit]],
         col = 'cornflowerblue', type = 'h', lwd = 2.0)
  
  axis(4)
  axis(3, mZ[yIdx[yHit]],  names(yHit[yHit]))
  axis(2)
  axis(1, mZ[bIdx[bHit]],  names(bHit[bHit]))

} 
```


```{r LKEATIQLDELNQK, fig.retina=3, fig.height=6, fig.width=8, fig.cap="LKEATIQLDELNQK Prosit predicted MS2; reproduce Fig.1d [@prosit2019] https://www.nature.com/articles/s41592-019-0426-7"}
.plotPrositIntensity("LKEATIQLDELNQK",
                     xlim = c(150, 1000),
                     ylim=c(0,1), axes=FALSE)
```

```{r Fig1, echo=FALSE, out.width="100%", eval=TRUE, fig.cap="Screenshot of Fig.1d [@prosit2019] https://www.nature.com/articles/s41592-019-0426-7"}
  knitr::include_graphics("https://user-images.githubusercontent.com/4901987/237583247-a948ceb3-b525-4c30-b701-218346a30cf6.jpg")
```


# Compare Intensity Predictions

## Get example data
Now we use `r BiocStyle::Biocpkg('tartar')` raw file as used in [@rawrr].

```{r ExperimentHub}
library(ExperimentHub)
eh <- ExperimentHub::ExperimentHub()
EH4547 <- normalizePath(eh[["EH4547"]])

(rawfile <- paste0(EH4547, ".raw"))

if (!file.exists(rawfile)){
	  file.copy(EH4547, rawfile)
}
```

## Plot

```{r Ms2GenericEnsemble}
i <- c(9594, 12677, 16136, 17193, 11113, 12788, 13204,
       11884, 13868, 14551, 17612)
S <- rawrr::readSpectrum(rawfile = rawfile, scan = i)

iRTpeptides <- c("LGGNEQVTR", "YILAGVENSK", "GTFIIDPGGVIR", "GTFIIDPAAVIR",
                 "GAGSSEPVTGLDAK", "TPVISGGPYEYR", "VEATFGVDESNAK",
                 "TPVITGAPYEYR", "DGLDAASYYAPVR", "ADVTPADFSEWSK",
                 "LFLQFGAQGSPFLK")

sapply(c(1:11), function(idx){
  massError <- (abs(((protViz::parentIonMass(iRTpeptides[idx]) + 1.007)/ 2) - S[[idx]]$pepmass))
  
  HCDEnergy <- S[[idx]]$`HCD Energy eV:` |> as.numeric()
  
  .plotPrositIntensity(iRTpeptides[idx],
                       xlim = c(150, 1000),
                       ylim=c(-1, 1),
                       sub = paste0(iRTpeptides[idx], " | HCD Energy eV:", HCDEnergy),
                       axes = FALSE,
                       collisionEnergy = 35,
                       precursorCharge = 2
  )
  
  points(S[[idx]]$mZ, -S[[idx]]$intensity / max(S[[idx]]$intensity), type = 'h')
  massError
})
```
## Collision Energy Normalization

```{r CollisionEnergyNormalization, fig.height=12}

## determine spectrum match
.spectrumMatch <- function(x0, y0, x1, y1){
  
  idx <- protViz::findNN(x0, x1)
  
  hits <- abs(x0 - x1[idx]) < 0.1
  fit <- lm(y1[idx[hits]] ~ y0[hits])
  
  #plot(y0[hits], y1[idx[hits]], sub=summary(fit)$r.squared)
  #abline(fit)

  summary(fit)$r.squared
}

.CENormalization <- function(peptide = "DGLDAASYYAPVR",
                             spectrum = NULL,
                             n = 30){

  peptides <- rep(peptide[1], n)
  collisionEnergies <- seq(10, 40, length = n)
  precursorCharges <- rep(2, n)
  ps <- koina::prosit2019Intensity(peptides, collisionEnergy = collisionEnergies, precursorCharge =  precursorCharges, usessl = TRUE) |>
    lapply(function(x)x[!is.na(x$mz),]) |>
    lapply(function(x) x[order(x$mz),])
  
  r.squared <- ps |>
    sapply(function(x) .spectrumMatch(x$mz, x$intensity, spectrum$mZ, spectrum$intensity))
  
  df <- data.frame(collisionEnergies = collisionEnergies,
                   r.squared = r.squared)
  
  plot(df,
       ylab='r.squared of matching peak intensities',
       main = peptide,
       sub = paste0("HCD Energy = ", spectrum$`HCD Energy:`),
       pch = 16, col='#99999999')
  abline(v = spectrum$`HCD Energy:`)
}

par(mfrow = c(4, 3))
rv <- lapply(1:11, function(i){
  .CENormalization(peptide = iRTpeptides[i], spectrum = S[[i]], n = 100) 
})

```

## Compare iRT Predictions

```{r iRTGenericEnsemble, fig.retina=3, fig.width=8, fig.height=8, eval=FALSE}
## compose data frame
rtdf <- data.frame(row.names = iRTpeptides,
                   Prosit = sapply(iRTpeptides, koina::prosit2019IrtEnsemble),
                   SSRC = protViz::ssrc(iRTpeptides),
                   iRT = sapply(iRTpeptides, function(x){protViz::iRTpeptides$rt[which(x == protViz::iRTpeptides$peptide)]}),
                   raw = sapply(S, function(x) x$rtinseconds)
)

.upper <- function(x, y){
  points(x, y, pch=16, col="#1155FF", cex=1.5)
  fit <- lm(y ~ x)
  abline(fit, col='#FFAAAAAA')
}

.lower <- function(x, y){
  fit <- lm(y ~ x)
  par(usr = c(0, 1, 0, 1))
  
  text(0.5, 0.5, paste0("r.squared=",
  round(summary(fit)$r.squared,3)),
  cex = 1.5)
}
pairs(rtdf, upper.panel = .upper, lower.panel = .lower)
```


# Generate a spec library


as provided by FragPipe

|  column| name                               | example            |
|--------|------------------------------------|--------------------|
|     1  |  PrecursorMz                       |  391.536391        |
|     2  |  ProductMz                         |  147.112805        |
|     3  |  Annotation                        |  y1^1              |
|     4  |  ProteinId                         |  P21333            |
|     5  |  GeneName                          |  FLNA              |
|     6  |  PeptideSequence                   |  DLAEDAPWKK        |
|     7  |  ModifiedPeptideSequence           |  DLAEDAPWKK        |
|     8  |  PrecursorCharge                   |  3                 |
|     9  |  LibraryIntensity                  |  10000.0           |
|    10  |  NormalizedRetentionTime           |  19.587033313296157|
|    11  |  PrecursorIonMobility              |                    |
|    12  |  FragmentType                      |  y                 |
|    13  |  FragmentCharge                    |  1                 |
|    14  |  FragmentSeriesNumber              |  1                 |
|    15  |  FragmentLossType                  |                    |
|    16  |  AverageExperimentalRetentionTime  |  1925.1328204860667|




```{r  eval=TRUE}
p <- koina::prosit2019Intensity(peptide = "LKEATIQLDELNQK",
                                 collisionEnergy = 35,
                                 precursorCharge = 3, usessl = TRUE)[[1]]


L <- koina:::.composeSpecLibrary(peptide = "LKEATIQLDELNQK",
                             proteinId = NA, collisionEnergy = 35,
                             precursorCharge = 3, p)
L |>
  knitr::kable(font_size = 7)

#knitr::kable(data.frame(name=names(L), full.name=cnames))
```

# FAQ

## Howto encode a modification?
```{r diffMod, fig.height=8, eval=FALSE}
x <- koina::prosit2019Intensity(peptide = "GDLLQVM[UNIMOD:35]HEAFEK",
                                 collisionEnergy = 35,
                                 precursorCharge = 3)[[1]]

y <- koina::prosit2019Intensity(peptide = "GDLLQVMHEAFEK",
                                 collisionEnergy = 35,
                                 precursorCharge = 3)[[1]]



x |> head()
 
y |> head()
```

# Session information {-}

```{r sessioninfo, echo=FALSE}
sessionInfo()
```


# References {-}

